drivers/adc/ad738x/iio_ad738x.c

```diff
@@ -0,0 +1,210 @@
+********************************************************************************
+* Copyright (c) 2024 Analog Devices, Inc.
+* Copyright (c) 2024 BayLibre, SAS.
+* All rights reserved.
+*
+* This software is proprietary to Analog Devices, Inc. and its licensors.
+* By using this software you agree to the terms of the associated
+* Analog Devices Software License Agreement.
+*******************************************************************************/
+#ifdef IIO_SUPPORT
+
+#include <stdio.h>
+#include <math.h>
+
+#include "ad738x.h"
+#include "iio_ad738x.h"
+#include "iio.h"
+#include "no_os_util.h"
+#include "no_os_alloc.h"
+#include "no_os_error.h"
+
+#define AD738X_CH(_idx) {\
+	.name = "voltage" #_idx, \
+	.ch_type = IIO_VOLTAGE,\
+	.ch_out = IIO_DIRECTION_INPUT,\
+	.indexed = true,\
+	.channel = _idx,\
+	.scan_index = _idx,\
+	.scan_type = &ad738x_iio_scan_type,\
+	.attributes = ad738x_iio_ch_attributes,\
+}
+
+#define AD738X_CHN_STORAGE_BITS	16
+#define AD738X_ADC_RESOLUTION		16
+#define AD738X_ADC_MAX_COUNT_UNIPOLAR  (uint32_t)((1 << AD738X_ADC_RESOLUTION) - 1)
+
+struct ad738x_iio {
+	struct ad738x_dev *drv_dev;
+	struct iio_device *iio_dev;
+};
+
+/**
+ * @brief Read the debug register value
+ * @param device - Pointer to IIO device instance
+ * @param reg - Register address to read from
+ * @param val - Pointer to variable to read data into
+ * @return 0 in case of success, negative value otherwise
+ */
+static int32_t iio_ad738x_debug_reg_read(void *device, uint32_t reg,
+		uint32_t *val)
+{
+	struct ad738x_iio_dev *iio_dev = device;
+	struct ad738x_dev *dev = iio_dev->ad738x_dev;
+	int ret;
+	uint16_t val16;
+
+	ret = ad738x_spi_reg_read(dev,  (uint8_t) reg, &val16);
+	if (ret)
+		return ret;
+
+	*val = val16;
+	return 0;
+}
+
+/**
+ * @brief Write the debug register value
+ * @param device - Pointer to IIO device instance
+ * @param reg - Register address to read from
+ * @param val - Pointer to variable to write data into
+ * @return 0 in case of success, negative value otherwise
+ */
+static int32_t iio_ad738x_debug_reg_write(void *device, uint32_t reg,
+		uint32_t val)
+{
+	struct ad738x_iio_dev *iio_dev = device;
+	struct ad738x_dev *dev = iio_dev->ad738x_dev;
+
+	return ad738x_spi_reg_write(dev, (uint8_t)reg, (uint16_t)val);
+}
+
+static int get_raw(void *device, char *buf, uint32_t len,
+		   const struct iio_ch_info *channel, intptr_t id)
+{
+	struct ad738x_iio_dev *iio_dev = device;
+	struct ad738x_dev *dev = iio_dev->ad738x_dev;
+	uint16_t results[2];
+	int32_t ret;
+
+	ret = ad738x_spi_single_conversion(dev, results);
+	if (ret)
+		return ret;
+
+	return sprintf(buf, "%i", results[channel->ch_num]);
+}
+
+static int get_scale(void *device, char *buf, uint32_t len,
+		     const struct iio_ch_info *channel, intptr_t id)
+{
+	struct ad738x_iio_dev *iio_dev = device;
+	struct ad738x_dev *dev = iio_dev->ad738x_dev;
+
+	return snprintf(buf, len, "%10f",
+			((double)dev->ref_voltage / AD738X_ADC_MAX_COUNT_UNIPOLAR));
```

Comment:
In general, we do not make floating point operations... there are other ways to return the scale in a string format using iio_format_value using only integers.
you may take a look at other projects which are using iio_format_value for scale and you may also take a look at the ad738x linux driver to see how the scale is computes.